### PR TITLE
[tests] Refactor importmulti tests

### PR DESCRIPTION
**Pull request by jnewbery**
https://github.com/bitcoin/bitcoin/pull/14565 needs test coverage. This PR refactors wallet_importmulti.py to the following pattern:

1. Add `get_key()` and `get_multisig()` methods, which generate keys on node0 and return the priv/pubkeys and all scriptPubKey and address variants.
2. Add `test_importmulti()` method, which takes an importmulti request, sends it to node1 and tests against success and error codes/messages.
3. Add `test_address()` method, which takes an address, sends it as a getaddressinfo request to node1 and tests the values returned.

This does not add any specific testing for #14565, but makes it very straightforward to add that testing: `test_importmulti()` can be easily updated to test for returned warnings, and `test_address()` can be called multiple times against the different address variants for a singlesig/multisig.

### CODE REVIEW

**Comment by jnewbery** [comment]
I've tidied up the intermediate commits and force-pushed. Should be ready for review.

**Comment by MeshCollider** [comment]
Concept ACK, will review very soon

File: test/functional/wallet_importmulti.py [hunk 1]
```diff
@@ -32,595 +80,523 @@ def skip_test_if_missing_module(self):
...
+        if error_code is not None:
+            assert_equal(result[0]['error']['code'], error_code)
+            assert_equal(result[0]['error']['message'], error_message)
+
+    def test_address(self, address, **kwargs):
+        """Get address info for `address` and test whether the returned values are as expected."""
+        addr_info = self.nodes[1].getaddressinfo(address)
+        for key, value in kwargs.items():
+            if addr_info[key] != value:
+                raise(AssertionError("key {} value {} did not match expected value {}".format(key, addr_info[key], value)))
```
Comment (practicalswift) [inline test/functional/wallet_importmulti.py]: Nit: `raise AssertionError(…)` is more idiomatic and consistent with the rest of the code base. Drop the unnecessary parens :-)

File: test/functional/wallet_importmulti.py [hunk 2]
```diff
@@ -32,595 +80,523 @@ def skip_test_if_missing_module(self):
...
+                   CScript([OP_0, pkh]).hex(),  # p2sh-p2wpkh redeem script
+                   key_to_p2sh_p2wpkh(pubkey))  # p2sh-p2wpkh addr
+
+    def get_multisig(self):
+        """Generate a fresh multisig on node0
+
+        Returns a named tuple of privkeys, pubkeys and all address and scripts."""
+        addrs = []
+        pubkeys = []
+        for i in range(3):
```
Comment (practicalswift) [inline test/functional/wallet_importmulti.py]: Nit: Use `_` instead of `i` to show that the variable is intentionally unused.

**Comment by jnewbery** [comment]
@practicalswift nits fixed and redundant code removed: https://github.com/bitcoin/bitcoin/compare/6c76af02e8188d0dc66c315e4ecd80308e2fa0a1..12e59a60019732fa228872848121cca4431e7513

File: test/functional/wallet_importmulti.py [hunk 3]
```diff
@@ -2,7 +2,18 @@
...
-"""Test the importmulti RPC."""
+"""Test the importmulti RPC.
+
+Test importmulti by generating keys on node0, importing the scriptPubKeys and
+addresses on node1 and then testing the address info for the different address
+variants.
+
+- `get_key()` and `get_multisig()` are called to generate keys on node0 and
+  return the privkeys, pubkeys and all variants of scriptPubKey and address.
+- `test_importmulti() is called to send an importmulti call to node1, test
```
Comment (MeshCollider) [inline test/functional/wallet_importmulti.py]: missing `

File: test/functional/wallet_importmulti.py [hunk 4]
```diff
@@ -209,10 +215,10 @@ def run_test(self):
                                "pubkeys": [key.pubkey],
                                "internal": False},
                               True)
-        address_assert = self.nodes[1].getaddressinfo(address)
-        assert_equal(address_assert['iswatchonly'], True)
-        assert_equal(address_assert['ismine'], False)
-        assert_equal(address_assert['timestamp'], timestamp)
+        self.test_address(address,
```
Comment (MeshCollider) [inline test/functional/wallet_importmulti.py]: This should also test for solvability

File: test/functional/wallet_importmulti.py [hunk 5]
```diff
@@ -511,18 +509,18 @@ def run_test(self):
...
         # P2WSH multisig address without scripts or keys
         multisig = self.get_multisig()
         self.log.info("Should import a p2wsh multisig as watch only without respective redeem script and private keys")
         self.test_importmulti({"scriptPubKey": {"address": multisig.p2wsh_addr},
                                "timestamp": "now"},
                               True)
-        address_assert = self.nodes[1].getaddressinfo(multisig.p2sh_addr)
-        assert_equal(address_assert['solvable'], False)
+        self.test_address(multisig.p2sh_addr,
+                          solvable=False)
```
Comment (MeshCollider) [inline test/functional/wallet_importmulti.py]: This should test watchonly=True

File: test/functional/wallet_importmulti.py [hunk 6]
```diff
@@ -304,10 +308,9 @@ def run_test(self):
                               False,
                               error_code=-8,
                               error_message='Internal must be set to true for nonstandard scriptPubKey imports.')
-        address_assert = self.nodes[1].getaddressinfo(address)
-        assert_equal(address_assert['iswatchonly'], False)
-        assert_equal(address_assert['ismine'], False)
-        assert_equal('timestamp' in address_assert, False)
```
Comment (MeshCollider) [inline test/functional/wallet_importmulti.py]: n.b. we've lost this test now, but that seems fine

File: test/functional/wallet_importmulti.py [hunk 7]
```diff
@@ -340,8 +343,7 @@ def run_test(self):
                                "timestamp": "now",
                                "redeemscript": multisig.redeem_script},
                               True)
-        address_assert = self.nodes[1].getaddressinfo(multisig.p2sh_addr)
-        assert_equal(address_assert['timestamp'], timestamp)
+        self.test_address(multisig.p2sh_addr, timestamp=timestamp)
```
Comment (MeshCollider) [inline test/functional/wallet_importmulti.py]: IMO it wouldn't hurt to test for solvability here too to ensure the same output as listunspent below

File: test/functional/wallet_importmulti.py [hunk 8]
```diff
@@ -360,8 +362,8 @@ def run_test(self):
                                "redeemscript": multisig.redeem_script,
                                "keys": multisig.privkeys[0:2]},
                               True)
-        address_assert = self.nodes[1].getaddressinfo(multisig.p2sh_addr)
-        assert_equal(address_assert['timestamp'], timestamp)
+        self.test_address(multisig.p2sh_addr,
+                          timestamp=timestamp)
```
Comment (MeshCollider) [inline test/functional/wallet_importmulti.py]: Same here

File: test/functional/wallet_importmulti.py [hunk 9]
```diff
@@ -579,8 +577,8 @@ def run_test(self):
                                "redeemscript": multisig.p2wsh_script,
                                "witnessscript": multisig.redeem_script},
                               True)
-        address_assert = self.nodes[1].getaddressinfo(multisig.p2sh_addr)
-        assert_equal(address_assert['solvable'], True)
+        self.test_address(address,
+                          solvable=True)
```
Comment (MeshCollider) [inline test/functional/wallet_importmulti.py]: test !ismine here?

**Review by meshcollider — COMMENTED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/14886/commits/12e59a60019732fa228872848121cca4431e7513, very nice, I've left a couple of minor comments inline but I assume you might want to keep this as a refactor rather than adding any extra testing here, so feel free to ignore the last few comments.

File: test/functional/wallet_importmulti.py [hunk 10]
```diff
@@ -55,20 +93,22 @@ def run_test(self):
 
         # Bitcoin Address (implicit non-internal)
         self.log.info("Should import an address")
-        address = self.nodes[0].getaddressinfo(self.nodes[0].getnewaddress())
+        key = self.get_key()
+        key = self.get_key()
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply test/functional/wallet_importmulti.py]: oops. Fixed intermediate commit.

File: test/functional/wallet_importmulti.py (hunk 3 above)
Comment (jnewbery) [reply test/functional/wallet_importmulti.py]: fixed

File: test/functional/wallet_importmulti.py [hunk 11]
```diff
@@ -84,24 +124,26 @@ def run_test(self):
...
         assert_equal(address_assert['iswatchonly'], True)
         assert_equal(address_assert['ismine'], False)
         assert_equal(address_assert['timestamp'], timestamp)
         assert_equal(address_assert['ischange'], True)
 
         # ScriptPubKey + internal + label
         self.log.info("Should not allow a label to be specified when internal is true")
-        address = self.nodes[0].getaddressinfo(self.nodes[0].getnewaddress())
+        key = self.get_key()
+        address = key.p2pkh_addr
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply test/functional/wallet_importmulti.py]: yes, removed

File: test/functional/wallet_importmulti.py (hunk 6 above)
Comment (jnewbery) [reply test/functional/wallet_importmulti.py]: I didn't think this test was important, so I removed it, but I've now readded it.

**Comment by jnewbery** [comment]
Thanks @MeshCollider . I've addressed your review comments and pushed: https://github.com/bitcoin/bitcoin/compare/12e59a60019732fa228872848121cca4431e7513..ee3b21dccbeb0f9d4e99de869dbfaf625c159c7f

> I've left a couple of minor comments inline but I assume you might want to keep this as a refactor rather than adding any extra testing here, so feel free to ignore the last few comments.

Yes, the intent was for this to be a straight refactor. I think more comprehensive testing can be added as part of #14565. In fact, I've started doing that here: https://github.com/jnewbery/bitcoin/tree/pr14565.tests. We should go through that commit and make sure that it's comprehensively testing all return values.

**Review by ryanofsky — APPROVED** [review]
utACK ee3b21dccbeb0f9d4e99de869dbfaf625c159c7f. This is a test-only change and nice cleanup, and I think it should be merged as is. There are more tests in John's branch that depend on this, and #14565 will depend on those tests, and there are several other prs based on #14565, so this should be given some priority.

File: test/functional/wallet_importmulti.py [hunk 12]
```diff
@@ -108,6 +108,14 @@ def get_multisig(self):
                         CScript([OP_HASH160, witness_script, OP_EQUAL]).hex(),  # p2sh-p2wsh
                         script_to_p2sh_p2wsh(script_code))  # p2sh-p2wsh addr
 
+    def test_importmulti(self, req, success, error_code=None, error_message=None):
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply test/functional/wallet_importmulti.py]: Agree, that at a minimum named arguments should be used at the caller.

**Comment by Sjors** [comment]
Post merge ACK

**Comment by jnewbery** [comment]
Thanks all for the review.

@ryanofsky - your review comments are fixed here: https://github.com/bitcoin/bitcoin/pull/14952

**Pull request merged**